docs/odcs.rst

```diff
@@ -24,84 +24,188 @@ idea is to not require the user to provide a yum repo url when building
+used. Continuing with our previous signing intent example, let's say a container
+image build request uses two composes. Compose 1 was generated with no signing
+keys provided, and compose 2 was generated with "my-release-key". In this case,
+the intent is "unsigned".
+
+Because compose IDs can be passed in to OSBS, a mechanism will be in place to
+classify the intent of an existing compose. This will be done by inspecting the
+signing keys used for generating the compose and performing a reverse mapping to
+determine the intent. If a match cannot be determined, intent will be set to
+``None``, which is less restrictive than any defined signing intents.
+
+The ``signing_intent`` specified in ``container.yaml`` can be overridden with
+the build parameter of same name. This particular parameter will be ignored for
+autorebuilds. The value in ``container.yaml`` should always be used in that
+case. Note that the signing intent used by the compose of parent image is still
+enforced.
+
+Enabling ODCS Integration
+"""""""""""""""""""""""""
+
+The first step in enabling ODCS integration is by configuring the OSBS
+environment to use a specific ODCS server via the ``odcs_*`` configuration
+parameters.
+
+Next, ``container.yaml`` must be created and it must include the *compose*
+section. See example further down. This file must then be committed and pushed
+to git repository. Simply building the container image as usual will trigger
+ODCS interaction.
+
+
+ODCS Compose Checks
+"""""""""""""""""""
+
+When an ODCS compose is passed in as an argument, OSBS will perform some checks
+to ensure the compose is usable.
+
+- State done: verify the compose is not about to expire, extending if needed.
+  Continue with OSBS build once compose is in this state.
+
+- State failed: not supported, causes failure of the OSBS build.
+
+- State removed: request compose to be regenerated. In the future, we may want
+  to check the ``pre-remove`` state: https://pagure.io/odcs/issue/90
+
+- State wait|generating: wait for a certain amount of time (configurable in OSBS
+  environment) for compose request to move out of this state. Polling is
+  performed during wait period.
+
 API Changes
 -----------
 
-The following parameters will be used by ``buildContainer`` task defined by
-``koji-containerbuild`` plugin:
+- ``signing_intent <str, default None>``
+
+  Exposed in `koji-containerbuild`_, and `osbs-client`_.
+
+  May not be used if ODCS integration is not enabled.
+
+  Overwrites ``signing_intent`` in ``container.yaml``.
+
+- ``compose_ids <list of ints>``
+
+  Exposed in `koji-containerbuild`_, and `osbs-client`_.
+
+  May not be used with ``yum_repourls`` parameter.
+
+  May not be used if ODCS integration is not enabled.
+
+  May not be used with ``signing_intent`` parameter.
+
+  Ignores *compose* section in ``container.yaml`` and does not request a new
+  ODCS compose to be created. The provided composes are used instead.
+
+
+CLI Changes
```

Comment:
Not sure if did understand correctly: will these CLI options override the values from container.yaml?

If so, shouldn't these be restricted to scratch builds only? (so that these values are always have a trace in dist-git for builds which can be potentially released)

Reply:
> will these CLI options override the values from container.yaml?

Yes. This is needed to give some flexibility when building a hierarchy of images without signed packages. The pattern I had considered was to have the most restrictive signing intent in `container.yaml`. In most cases, this would be "release", but for upcoming products this could be "beta". The CLI argument would allow users to downgrade this to "unsigned", for example. This is critical in allowing workflows for daily builds. Currently, initiating signing of packages is done via a manual process which is somewhat time consuming.

I hadn't considered the scenario of upgrading the signing intent. For instance, if `container.yaml` defines "beta", but "release" is used in the CLI. I guess in this case it's fine. If build is successful, it really means that packages used have been signed with the "release" intent. This value is also recorded in the Koji build metadata for tracking purposes.

Reply:
I'm also a bit wary of not also requiring scratch builds when setting this values using the CLI. Compare with overriding architectures from the CLI, which cannot be done unless the build is a scratch build. Can we include some rationale on that?

Reply:
Added this:
> The Koji build metadata will contain a new key, build.extra.image.odcs.signing_intent_overridden, to indicate whether or not the signing_intent was overridden (CLI parameter, automatically downgraded, etc). This value will only be true if build.extra.image.odcs.signing_intent matches the signing_intent in container.yaml.